### ISSUE TITLE
`DataFrame.limit(n)` does not work when `n` is greater than table size

### ISSUE THREAD

**Issue by kevinzwang** [Eventual-Inc/Daft#2730]
**Describe the bug**
When a DataFrame has a micropartition that contains multiple tables and you do `df.limit` with a number greater than or equal to the size of the first table, it will not actually reduce the amount of rows correctly and will just get the entire micropartition.

**To Reproduce**
1. Download and unzip this file: [test_data.zip](https://github.com/user-attachments/files/16753745/test_data.zip). It contains a directory `test_data/` with two files, `1.parquet` and `2.parquet`, both containing four rows of an Int64 column `a`.
2. In Python, run
```python
>>> import daft
>>> df = daft.read_parquet("test_data/*") # will read it into one micropartition with two tables
>>> df.show()                                                                                                                                                                                                                                                                                        ╭───────╮                                                                                                                                                                                                                                                                                                       
│ a     │
│ ---   │
│ Int64 │
╞═══════╡
│ 1     │
├╌╌╌╌╌╌╌┤
│ 2     │
├╌╌╌╌╌╌╌┤
│ 3     │
├╌╌╌╌╌╌╌┤
│ 4     │
├╌╌╌╌╌╌╌┤
│ 5     │
├╌╌╌╌╌╌╌┤
│ 6     │
├╌╌╌╌╌╌╌┤
│ 7     │
├╌╌╌╌╌╌╌┤
│ 8     │
╰───────╯
Table with 8 rows and 1 columns

(Showing first 8 rows)
>>> df.limit(2).show() # or any limit between 1 and 7
[output 1]
>>> df.collect().limit(5).show() # or any limit between 4 and 7
[output 2]
```

**Expected behavior**
Both should show a dataframe with the correct number of rows.
output 1:
```
╭───────╮                                                                                                                                                                                                                                                                                                      
│ a     │                                                                                                                                                                                                                                                                                                      
│ ---   │
│ Int64 │
╞═══════╡
│ 1     │
├╌╌╌╌╌╌╌┤
│ 2     │
╰───────╯
Table with 2 rows and 1 columns

(Showing first 2 of 2 rows)
```
output 2:
```
╭───────╮                                                                                                                                                                                                                                                                                                      
│ a     │
│ ---   │
│ Int64 │
╞═══════╡
│ 1     │
├╌╌╌╌╌╌╌┤
│ 2     │
├╌╌╌╌╌╌╌┤
│ 3     │
├╌╌╌╌╌╌╌┤
│ 4     │
├╌╌╌╌╌╌╌┤
│ 5     │
╰───────╯
Table with 5 rows and 1 columns

(Showing first 5 of 5 rows)

```

**Actual Behavior**
For output 1, if you do not materialize before the limit, it returns at most `n*2` rows, `n` from each file.
```python
>>> df.limit(2).show()
╭───────╮                                                                                                                                                                                                                                                                                                      
│ a     │                                                                                                                                                                                                                                                                                                      
│ ---   │
│ Int64 │
╞═══════╡
│ 1     │
├╌╌╌╌╌╌╌┤
│ 2     │
├╌╌╌╌╌╌╌┤
│ 5     │
├╌╌╌╌╌╌╌┤
│ 6     │
╰───────╯
Table with 4 rows and 1 columns

(Showing first 4 of 4 rows)
```
For output 2, if the data is materialized before the limit, if `n` is greater than or equal to the length of the first table (4), it will return the entire dataframe.
```python
>>> df.collect().limit(5).show()
╭───────╮                                                                                                                                                                                                                                                                                                      
│ a     │
│ ---   │
│ Int64 │
╞═══════╡
│ 1     │
├╌╌╌╌╌╌╌┤
│ 2     │
├╌╌╌╌╌╌╌┤
│ 3     │
├╌╌╌╌╌╌╌┤
│ 4     │
├╌╌╌╌╌╌╌┤
│ 5     │
├╌╌╌╌╌╌╌┤
│ 6     │
├╌╌╌╌╌╌╌┤
│ 7     │
├╌╌╌╌╌╌╌┤
│ 8     │
╰───────╯
Table with 8 rows and 1 columns

(Showing first 8 rows)
```

This isn't an issue with `show`, it will do the same thing for `collect` or `to_pandas` as well.

**Desktop (please complete the following information):**
 - OS: MacOS, Linux
 - Version 0.3.1

**Additional Context**
```python
>>> df.limit(3).explain(show_all=True)
== Unoptimized Logical Plan ==

* Limit: 3
|
* GlobScanOperator
|   Glob paths = [test_data/*]
|   Coerce int96 timestamp unit = Nanoseconds
|   IO config = S3 config = { Max connections = 8, Retry initial backoff ms = 1000, Connect timeout ms = 30000, Read timeout ms = 30000, Max retries = 25, Retry mode = adaptive, Anonymous = false, Use SSL = true, Verify SSL = true, Check hostname SSL = true, Requester pays = false, Force Virtual
|     Addressing = false }, Azure config = { Use Fabric Endpoint = false, Anonymous = false, Use SSL = true }, GCS config = { Anonymous = false }, HTTP config = { user_agent = daft/0.0.1 }
|   Use multithreading = true
|   File schema = a#Int64
|   Partitioning keys = []
|   Output schema = a#Int64


== Optimized Logical Plan ==

* Limit: 3
|
* GlobScanOperator
|   Glob paths = [test_data/*]
|   Coerce int96 timestamp unit = Nanoseconds
|   IO config = S3 config = { Max connections = 8, Retry initial backoff ms = 1000, Connect timeout ms = 30000, Read timeout ms = 30000, Max retries = 25, Retry mode = adaptive, Anonymous = false, Use SSL = true, Verify SSL = true, Check hostname SSL = true, Requester pays = false, Force Virtual
|     Addressing = false }, Azure config = { Use Fabric Endpoint = false, Anonymous = false, Use SSL = true }, GCS config = { Anonymous = false }, HTTP config = { user_agent = daft/0.0.1 }
|   Use multithreading = true
|   File schema = a#Int64
|   Partitioning keys = []
|   Limit pushdown = 3
|   Output schema = a#Int64


== Physical Plan ==

* Limit: 3
|   Eager = false
|   Num partitions = 1
|
* TabularScan:
|   Num Scan Tasks = 1
|   Estimated Scan Bytes = 1170
|   Clustering spec = { Num partitions = 1 }
|   Pushdowns: {limit: 3}
|   Schema: {a#Int64}
|   Scan Tasks: [
|   {File {file://test_data/1.parquet}, File {file://test_data/2.parquet}}
|   ]

>>> df.collect().limit(5).explain(show_all=True)
== Unoptimized Logical Plan ==                                                                                                                                                                                                                                                                                 

* Limit: 5
|
* Source:
|   Number of partitions = 1
|   Output schema = a#Int64


== Optimized Logical Plan ==

* Limit: 5
|
* Source:
|   Number of partitions = 1
|   Output schema = a#Int64


== Physical Plan ==

* Limit: 5
|   Eager = false
|   Num partitions = 1
|
* InMemoryScan:
|   Schema = a#Int64,
|   Size bytes = 1404,
|   Clustering spec = { Num partitions = 1 }
```

**Comment by kevinzwang** [comment]
Ah figured it out. PR out soon

**Comment by kevinzwang** [comment]
Oh this was actually fixed in 0.3.1. Thanks @colin-ho! I am running into an unrelated limit issue with my TPC-H benchmarking then... will try to repro with the correct version
